src/hyperlight_host/src/hypervisor/kvm.rs

```diff
@@ -55,6 +65,527 @@ pub(crate) fn is_hypervisor_present() -> bool {
+
+                data = &data[write_len..];
+                gva += write_len as u64;
+            }
+
+            Ok(())
+        }
+
+        pub fn read_regs(&self, regs: &mut X86_64Regs) -> Result<()> {
+            log::debug!("Read registers");
+            let vcpu_regs = self
+                .vcpu_fd
+                .get_regs()
+                .map_err(|e| new_error!("Could not read guest registers: {:?}", e))?;
+
+            regs.rax = vcpu_regs.rax;
+            regs.rbx = vcpu_regs.rbx;
+            regs.rcx = vcpu_regs.rcx;
+            regs.rdx = vcpu_regs.rdx;
+            regs.rsi = vcpu_regs.rsi;
+            regs.rdi = vcpu_regs.rdi;
+            regs.rbp = vcpu_regs.rbp;
+            regs.rsp = vcpu_regs.rsp;
+            regs.r8 = vcpu_regs.r8;
+            regs.r9 = vcpu_regs.r9;
+            regs.r10 = vcpu_regs.r10;
+            regs.r11 = vcpu_regs.r11;
+            regs.r12 = vcpu_regs.r12;
+            regs.r13 = vcpu_regs.r13;
+            regs.r14 = vcpu_regs.r14;
+            regs.r15 = vcpu_regs.r15;
+
+            regs.rip = vcpu_regs.rip;
+            regs.rflags = vcpu_regs.rflags;
+
+            Ok(())
+        }
+
+        pub fn write_regs(&self, regs: &X86_64Regs) -> Result<()> {
+            log::debug!("Write registers");
+            let new_regs = kvm_regs {
+                rax: regs.rax,
+                rbx: regs.rbx,
+                rcx: regs.rcx,
+                rdx: regs.rdx,
+                rsi: regs.rsi,
+                rdi: regs.rdi,
+                rbp: regs.rbp,
+                rsp: regs.rsp,
+                r8: regs.r8,
+                r9: regs.r9,
+                r10: regs.r10,
+                r11: regs.r11,
+                r12: regs.r12,
+                r13: regs.r13,
+                r14: regs.r14,
+                r15: regs.r15,
+
+                rip: regs.rip,
+                rflags: regs.rflags,
+            };
+
+            self.vcpu_fd
+                .set_regs(&new_regs)
+                .map_err(|e| new_error!("Could not write guest registers: {:?}", e))
+        }
+
+        pub fn add_hw_breakpoint(&mut self, addr: u64) -> Result<bool> {
+            let addr = self.translate_gva(addr)?;
+
+            if let Some(debug) = self.debug.as_mut() {
+                if debug.hw_breakpoints.contains(&addr) {
+                    Ok(true)
+                } else if debug.hw_breakpoints.len() >= KvmDebug::MAX_NO_OF_HW_BP {
+                    Ok(false)
+                } else {
+                    debug.hw_breakpoints.push(addr);
+                    debug.set_breakpoints(&self.vcpu_fd, false)?;
+
+                    Ok(true)
+                }
+            } else {
+                Ok(false)
+            }
+        }
+
+        pub fn remove_hw_breakpoint(&mut self, addr: u64) -> Result<bool> {
+            let addr = self.translate_gva(addr)?;
+
+            if let Some(debug) = self.debug.as_mut() {
+                if debug.hw_breakpoints.contains(&addr) {
```

Comment:
Seems like the `.contains` is unnecessary given that you use `.position` later anyway. You might be able to simplify this whole block using `hw_breakpoints.retain(|a| a != addrv)` or something along those lines

Reply:
thanks for the suggestion